### PR TITLE
Add User Toggle

### LINKED ISSUE #72: Show / Hide Recommendations View
I'd like the ability to show / hide recommendations.

Example, when I switch to grid view I find the Recommendations visually distracting.

- [ ] Requires https://github.com/nextcloud/server/issues/16602

### PR DESCRIPTION
**Pull request by gary-kim**
This adds the ability for users to disable the app for themselves from settings.

Closes #72

### CODE REVIEW

File: js/personal-settings.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+$(document).ready(() => {
```
Comment (ChristophWurst) [inline js/personal-settings.js]: Please don't use the global `$` but add another webpack entry point for this https://blog.wuc.me/2019/06/12/problem-nextcloud-global-dependencies.html

File: js/personal-settings.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+	$('#recommendations-enabled').on('click', () => {
+		let recommendationsEnabled = document.getElementById('recommendations-enabled');
+		$.ajax({
```
Comment (ChristophWurst) [inline js/personal-settings.js]: use [Axios](https://github.com/nextcloud/recommendations/blob/154da86a2b78fd5bd2bb46f8aa47ea7d9a1b58eb/src/service/RecommendationService.js#L28)

File: lib/Controller/SettingsController.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	 * @return JSONResponse
+	 */
+	public function getSettings (): JSONResponse {
+		$user = $this->userSession->getUser();
+		if (is_null($user)) {
+			throw new Exception("Not logged in");
+		}
+		return new JSONResponse([
+			'status' => 'success',
+			'enabled' => $this->config->getUserValue($user->getUID(), Application::APP_ID, 'enabled', 'true'),
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php]:
```suggestion
			'enabled' => $this->config->getUserValue($user->getUID(), Application::APP_ID, 'enabled', 'true') === true,
```

to make it a proper boolean value

File: lib/Controller/SettingsController.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	 * 
+	 * @return JSONResponse
+	 */
+	public function getSettings (): JSONResponse {
+		$user = $this->userSession->getUser();
+		if (is_null($user)) {
+			throw new Exception("Not logged in");
+		}
+		return new JSONResponse([
+			'status' => 'success',
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php]: 200 is always success. error should be transported via the HTTP status code

File: lib/Controller/SettingsController.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+			throw new Exception("Not logged in");
+		}
+
+		$avaliableSetttings = ['enabled'];
+
+		if (!in_array($key, $avaliableSetttings)) {
+			return new JSONResponse([
+				'status' => 'error',
+				'message' => 'parameter does not exist',
+			], Http::STATUS_BAD_REQUEST);
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php]: use `STATUS_UNPROCESSABLE_ENTITY`

File: lib/Controller/SettingsController.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		$user = $this->userSession->getUser();
+		if (is_null($user)) {
+			throw new Exception("Not logged in");
+		}
+
+		$avaliableSetttings = ['enabled'];
+
+		if (!in_array($key, $avaliableSetttings)) {
+			return new JSONResponse([
+				'status' => 'error',
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php]: remove status

File: lib/Controller/SettingsController.php [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		}
+
+		$param = $this->request->getParam('data');
+		if (!isset($param)) {
+			$param = 'false';
+		}
+		$this->config->setUserValue($user->getUID(), Application::APP_ID, $key, $param);
+
+		return new JSONResponse([
+			'status' => 'success',
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php]: remove status and return the newly set value

File: lib/Controller/SettingsController.php [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+		]);
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * 
+	 * @param $key
+	 * @return JSONResponse
+	 */
+	public function setSetting ($key): JSONResponse {
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php]: shouldn't it be `(string $key, string $value)`?

File: lib/Settings/Personal.php [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+		$this->user = $user;
+	}
+
+	/**
+	 * @return TemplateResponse
+	 */
+	public function getForm() {
+		$userId = $this->user->getUser()->getUID();
+		$parameters = [
+			'enabled' => $this->config->getUserValue($userId, Application::APP_ID, 'enabled', 'true'),
```
Comment (ChristophWurst) [inline lib/Settings/Personal.php]:
```suggestion
			'enabled' => $this->config->getUserValue($userId, Application::APP_ID, 'enabled', 'true') === 'true',
```

File: src/main.js [hunk 10]
```diff
@@ -23,40 +23,51 @@ import Vue from "vue";
...
-		const vm = new View({
-			propsData: {}
-		}).$mount(el);
-
-		fileList.$el.on('changeDirectory', data => {
-			if (data.dir.toString() === '/') {
-				vm.show();
-			} else {
-				vm.hide();
+		Axios.get(generateUrl('/apps/recommendations/settings')).then((res) => {
```
Comment (ChristophWurst) [inline src/main.js]: :-1: I don't like this as it causes two requests.

Why don't you just return an empty recommendations array when they are disabled? Then you don't need any changes here as an empty list isn't rendered anyway.

Comment (gary-kim) [reply src/main.js]: Can I merge in [this branch](https://github.com/gary-kim/recommendations/tree/add-toggle-controller) then? It changes the controller rather then the client side.

Comment (ChristophWurst) [reply src/main.js]: Yes, that looks better!

File: lib/Controller/SettingsController.php (hunk 8 above)
Comment (gary-kim) [reply lib/Controller/SettingsController.php]: I just tried that and got a null in the second argument rather then a string. The current implementation seems to work.

**Comment by gary-kim** [comment]
Thank you for your feedback @ChristophWurst.
I've made most of the changes you have suggested.

File: lib/Controller/SettingsController.php (hunk 8 above)
Comment (ChristophWurst) [reply lib/Controller/SettingsController.php]: I'm pretty sure the type hints are fine. If they aren't then the request does not send a `value` parameter. Have you adjusted the AJAX code accordingly?

File: package.json [hunk 11]
```diff
@@ -19,6 +19,7 @@
   },
   "homepage": "https://github.com/ChristophWurst/files_recommendation#readme",
   "dependencies": {
+    "jquery": "^3.4.1",
```
Comment (ChristophWurst) [inline package.json]: we'll find a way around jquery :P

File: src/personal-settings.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+import $ from 'jquery';
+import Axios from 'nextcloud-axios';
+
+$(document).ready(() => {
```
Comment (ChristophWurst) [inline src/personal-settings.js]: In general I would prefer a tiny Vue settings component. Do you have experience with Vue? It's really simple. You can check out other code like https://github.com/nextcloud/twofactor_u2f/tree/master/src. You can create a settings component like https://github.com/nextcloud/twofactor_u2f/blob/master/src/components/PersonalSettings.vue and initialize it like in https://github.com/nextcloud/twofactor_u2f/blob/master/src/main-settings.js. Just let me know if you need help :)

File: src/webpack.common.js [hunk 13]
```diff
@@ -2,7 +2,10 @@ const path = require('path');
 const { VueLoaderPlugin } = require('vue-loader');
 
 module.exports = {
-	entry: path.join(__dirname, 'main.js'),
+	entry: {
+		main: path.join(__dirname, 'main.js'),
+		'personal-settings': path.join(__dirname, 'personal-settings.js')
```
Comment (ChristophWurst) [inline src/webpack.common.js]: awesome!

File: lib/Controller/SettingsController.php (hunk 8 above)
Comment (gary-kim) [reply lib/Controller/SettingsController.php]: My bad, I didn't change the ajax code yet. I've made the change.

File: src/personal-settings.js (hunk 12 above)
Comment (gary-kim) [reply src/personal-settings.js]: Thanks for the references. I'll try to figure it out.

Comment (ChristophWurst) [reply src/personal-settings.js]: They are okay for simple tasks. But things usually just get more complex and the php templating does not scale.

File: package.json (hunk 11 above)
Comment (gary-kim) [reply package.json]: jQuery has been removed.

File: src/personal-settings.js (hunk 12 above)
Comment (gary-kim) [reply src/personal-settings.js]: Personal settings have been changed to use Vue.

File: lib/Controller/SettingsController.php [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+		$avaliableSetttings = ['enabled'];
+
+		if (!in_array($key, $avaliableSetttings)) {
+			return new JSONResponse([
+				'message' => 'parameter does not exist',
+			], Http::STATUS_UNPROCESSABLE_ENTITY);
+		}
+
+		if (is_null($value)) {
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php]: it can't be null now with the type hint :)

**Comment by gary-kim** [comment]
Sorry for the long back and forth @ChristophWurst. First large pull request for Nextcloud.
As an aside, could I add my name to the authors list?

**Comment by ChristophWurst** [comment]
> Sorry for the long back and forth @ChristophWurst. First large pull request for Nextcloud.

No problem!

> As an aside, could I add my name to the authors list?

In the license header? Sure, go ahead :)

**Comment by gary-kim** [comment]
> In the license header? Sure, go ahead :)

Sorry, I meant the `info.xml`.

Also, before merging, there's a bit of indentation issues I'd like to fix.

**Comment by jancborchardt** [comment]
Hi @gary-kim! :) Thanks for your contribution!

Is it also possible to put the checkbox in the user-facing settings in the bottom left of the Files app?
Show recommendations setting


It would be much simpler to have an entry "Show recommendations" in there (enabled by default) and not add more stuff to our settings. :)

**Comment by gary-kim** [comment]
> Is it also possible to put the checkbox in the user-facing settings in the bottom left of the Files app?

Thank you for your recommendation @jancborchardt.

I also wanted to put it in the sidebar at first but the sidebar settings are from the Files app and cannot be changed from another app. Perhaps in the long term, this app could be integrated into the Files app but until then, it needs a separate settings page.

In addition, there are proposals that would require adding more settings for Recommendations such as #78 which may make the options too complex for the sidebar.

**Comment by jancborchardt** [comment]
With Nextcloud we want to keep options low and to the necessary elements. If already a small feature like Recommendations has more settings than Files itself, we know there's a problem. ;)

Since the settings in the bottom left of Files come from the Files app: This change should be done in the server repo / files app instead, check if the Recommendations app is enabled, and in that case add the setting.

**Comment by gary-kim** [comment]
> check if the Recommendations app is enabled, and in that case add the setting.

I like that idea. I'll move what is currently in this pull request into another branch in case we ever need the settings page then work on adding the settings to the files app instead.

**Comment by jancborchardt** [comment]
@gary-kim I commented on your issue in server to approve it. :+1: @ChristophWurst has to comment on the technicalities here – but from the design side I don’t see the need for settings anytime soon.

**Comment by skjnldsv** [comment]
@jancborchardt well, we should not add code that concern another app into core as this will get forgotten and is a bad code design.
If a toggle should be added, this should be from this app. :)

**Comment by jancborchardt** [comment]
> and is a bad coding design

It's apparently not possible to inject a setting into Files, so that would then have to be implemented. Lots of overhead, but your call. :)

**Comment by gary-kim** [comment]
I would also prefer not to have code only for Recommendations as part of the core. Unfortunately, that would mean that Recommendations needs its own page in settings or the work needs to be done to give some way to inject settings into files.

As a third option, the admin settings have an `additional` section. Could we have something similar apps such as Recommendations declare and use?

**Comment by skjnldsv** [comment]
> I would also prefer not to have code only for Recommendations as part of the core. Unfortunately, that would mean that Recommendations needs its own page in settings or the work needs to be done to give some way to inject settings into files.

If that means having this to ensure we don't pollute core, then so be it. :)
@juliushaertl, thoughts?

**Comment by ChristophWurst** [comment]
IMO the only sensible options are

* Create a new API for the files app where apps can inject arbitrary settings features and use that with this app (very complex)
* Have a new settings section in the personal settings (simple and currently what this PR does. but of course not the best UX)

We should refrain any other "solutions". Files or server should not know about recommendations and therefore those settings must not be included there.

Therefore I would either do it like this PR does or not allow this feature at all.

**Comment by skjnldsv** [comment]
Let's open an issue on server to request a setting api for files
and merge this pr?

**Comment by gary-kim** [comment]
Sounds good.
I'll rebase this branch and close the issue on the server.

**Comment by ChristophWurst** [comment]
> EDIT: Oh no, master keeps changing! How do you rebase onto a moving base? :)

Sorry. I merged a few Dependabot bumps :)

> EDIT 2: Also, would this be worth backporting?

No, we don't backport features.

**Comment by jancborchardt** [comment]
> Let's open an issue on server to request a setting api for files
> and merge this pr?

@skjnldsv does this issue exist now? Because the current state is suboptimal with an entirely separate settings section just for Recommendations, it feels like an afterthought.

I’m not sure why we had to merge this so fast instead of going with https://github.com/nextcloud/server/issues/16155 ?

**Comment by jancborchardt** [comment]
Reopened https://github.com/nextcloud/server/issues/16155 to reflect that this stil needs work

**Comment by skjnldsv** [comment]
> @skjnldsv does this issue exist now? Because the current state is suboptimal with an entirely separate settings section just for Recommendations, it feels like an afterthought.
> 
> I’m not sure why we had to merge this so fast instead of going with [nextcloud/server#16155](https://github.com/nextcloud/server/issues/16155) ?

https://github.com/nextcloud/recommendations/pull/100#issuecomment-507023856


> It's apparently not possible to inject a setting into Files, so that would then have to be implemented. Lots of overhead, but your call. :)

https://github.com/nextcloud/recommendations/pull/100#issuecomment-507165256

Please read the previous answers :)

yes this is an over powered solution, but until we have a proper way to add entries in the files settings, we will not add this into core :hugs:

**Comment by skjnldsv** [comment]
> @skjnldsv does this issue exist now?

Not that I'm aware of :)

**Comment by jancborchardt** [comment]
> yes this is an over powered solution, but until we have a proper way to add entries in the files settings, we will not add this into core

Right – in that case I would say it would be better to not have the setting (also as @ChristophWurst said as another option)? That’s kinda what I wanted to imply with my previous comment, but I understand how it wasn’t clear. :)

**Comment by jancborchardt** [comment]
Also just to clarify: Yes we shouldn’t bloat the core. But now we’re in a situation where the UX of the default installation is off. From the user perspective, Recommendations are 100% part of Files, so that’s where as setting would be expected. :)

**Comment by skjnldsv** [comment]
> Yes we shouldn’t bloat the core. But

Sorry, there is nothing following this sentence that will be good :)
Yes ux speaking it make sense, no the current code state does not grant a proper implementation. I already said it in another ticket, but ux is not a dictator for bad coding habits just because it will look good. That's is how we got to the point of the current files app and core in general. We're losing hundred of work hours removing and cleaning stuff because no one cared to refrain themselves from implementing features for the sake of it.

Please either do it nicely or don't do it at all.
To quote a wise man:

> We should refrain any other "solutions". Files or server should not know about recommendations and therefore those settings must not be included there.

**Comment by jancborchardt** [comment]
> We're losing hundred of work hours removing and cleaning stuff because no one cared to refrain themselves from implementing features for the sake of it.
>  Please either do it nicely or don't do it at all.

Right, and this can 100% be applied to the UX part as well.

Hence, and I’m sorry to say @gary-kim, it might be best to revert this and work out how we can add this in the proper place instead.

This is my fault as I was on vacation for 2 weeks during my last comments above end of June and I should have said that the setting should go into Files or not at all. :(

**Comment by skjnldsv** [comment]
> This is my fault as I was on vacation for 2 weeks during my last comments above end of June and I should have said that the setting should go into Files or not at all. :(

No fault here :smile_cat: 
Just decisions to make! :wink: 

You're also 100% right on the ux part. Let's revert then! :muscle:

**Comment by jancborchardt** [comment]
Thank you @skjnldsv! 😊

Sorry again @gary-kim for not being clear! Btw, are you going to come to our Nextcloud Conference in Berlin in September? https://nextcloud.com/conf-2019/ for contributors it’s free and there’s travel & hotel support! Would be really great to have you. 🚀

**Comment by skjnldsv** [comment]
Revert in https://github.com/nextcloud/recommendations/pull/110

**Comment by GlenHuf** [comment]
Can't the "setting" go into the recommendations app itself?   just add a tinny toggle button or triangle to minimize or enable the recommendations pane.  That's more simple than a check box down in the lower left by user settings.

**Pull request merged**